app/api/v3/endpoints/rank_router.py

```
@@ -0,0 +1,21 @@
+        rank_service = RankServiceV3()
+        return rank_service.get_ranked_portfolios(req)
+    except Exception as e:
+        raise HTTPException(status_code=500, detail=str(e))
+
+@router.post("/portfolios/by-ad-elements", response_model=RankDTOV3.GetRankPtfoResponse)
+async def get_ranked_portfolios_by_ad_elements(req: RankDTOV3.GetRankPtfoByAdElementsRequest):
+    try:
+        rank_service = RankServiceV3()
+        return rank_service.get_ranked_portfolios_by_ad_elements(req)
+    except Exception as e:
+        raise HTTPException(status_code=500, detail=str(e))
```

Ruff (0.12.2)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)